SDK/MobileCenterPush/Microsoft.Azure.Mobile.Push.UWP/Push.cs

```diff
@@ -1,67 +1,174 @@
 using Windows.Networking.PushNotifications;
 
 namespace Microsoft.Azure.Mobile.Push
 {
+    using WindowsPushNotificationReceivedEventArgs = Windows.Networking.PushNotifications.PushNotificationReceivedEventArgs;
+
     public partial class Push : MobileCenterService
     {
-        // Retrieve the push token from platform-specific Push Notification Service,
-        // and later use the token to register with Mobile Center backend.
-        private void InstanceRegister()
-        {
-            if (!Enabled)
-            {
-                MobileCenterLog.Warn(MobileCenterLog.LogTag, "Push service is not enabled.");
-            }
+        private ApplicationLifecycleHelper _lifecycleHelper = new ApplicationLifecycleHelper();
 
-            _stateKeeper.InvalidateState();
-            var stateSnapshot = _stateKeeper.GetStateSnapshot();
-            _mutex.Unlock();
+        private PushNotificationChannel _channel;
 
-            var pushNotificationChannel = Task.Run(() => CreatePushNotificationChannel()).Result;
+        private static event EventHandler<PushNotificationReceivedEventArgs> PlatformPushNotificationReceived;
 
-            try
+        /// <summary>
+        /// Call this method at the end of Application.OnLaunched with the same parameter as OnLaunched.
+        /// This method call is needed to handle click on push to trigger the portable PushNotificationReceived event.
+        /// </summary>
+        /// <param name="e">OnLaunched method event</param>
+        public static void CheckPushActivation(LaunchActivatedEventArgs e)
+        {
+            if (PlatformPushNotificationReceived != null && Enabled)
             {
-                _mutex.Lock(stateSnapshot);
+                var customData = ParseLaunchString(e?.Arguments);
+                if (customData != null)
+                {
+                    PlatformPushNotificationReceived.Invoke(null, new PushNotificationReceivedEventArgs()
+                    {
+                        Title = null,
+                        Message = null,
+                        CustomData = customData
+                    });
+                }
             }
-            catch (StatefulMutexException e)
+        }
+
+        /// <summary>
+        /// If enabled, register push channel and send URI to backend.
+        /// Also start intercepting pushes.
+        /// If disabled and previously enabled, stop listening for pushes (they will still be received though).
+        /// </summary>
+        private void ApplyEnabledState()
+        {
+            // Since the lock we use is not recursive, caller of this method is expected to execute this method inside lock
+            if (Enabled)
             {
-                MobileCenterLog.Warn(MobileCenterLog.LogTag, "Push service registering with Mobile Center backend has failed", e);
-                return;
+                var stateSnapshot = _stateKeeper.GetStateSnapshot();
+                Task.Factory.StartNew(async () =>
```

Comment:
I think moving the body of this lambda into its own method would make this much more readable, but I won't press the issue either way